introduce sanity check when creating bufferdatasink

Gasoonjia opened the pull request:
Summary: This diff introduces sanity check on data alignment when creating buffer_data_sink. Also, for better control the raised error, we changed to use static function instead of constructor for constuction.

Differential Revision: D70190912

swolchok reviewed (approved):
seems reasonable

Gasoonjia commented:
@pytorchbot label "topic: not user facing"

dbort commented on devtools/etdump/tests/buffer_data_sink_test.cpp:
```diff
@@ -39,11 +39,13 @@ class BufferDataSinkTest : public ::testing::Test {
 TEST_F(BufferDataSinkTest, StorageSizeCheck) {
-  Result<size_t> ret = data_sink_->get_storage_size();
+  Result<BufferDataSink> buffer_data_sink = BufferDataSink::create(buffer_);
+  ASSERT_EQ(buffer_data_sink.error(), Error::Ok);
```
It'd be good to put this same ASSERT after all `create` calls in this file

Gasoonjia replied:
Thanks for the suggestion! Making `buffer_data_sink` as a private variable in `BufferDataSinkTest` class, and putting ASSERT after creations, for better structure.

dbort reviewed (approved):
This is a great change; a static factory is a much better way to handle creation in our exception-free system.

Gasoonjia committed:
introduce sanity check when creating bufferdatasink (#8709)

Summary:

This diff introduces sanity check on data alignment when creating buffer_data_sink. Also, for better control the raised error, we changed to use static function instead of constructor for constuction.

Reviewed By: swolchok

Differential Revision: D70190912
